### PR TITLE
CatDog: Add skin selector and add a new friend :^)

### PR DESCRIPTION
**Pull request by TheMorc**
~~I'm not sure how you guys will react.~~
~~But yes, I know, it might get a bit controversial, I think that our CatDog deserves a proper, better name, so let's call him Muf! _(Taking inspiration from my cat)_~~
~~I also reworked the image part so it uses one whole image PNG and not a bunch of small separate PNGs. It also saves us a few lines of code if that does even matter to someone. I would also want to implement a "skin selector" thing sometime later on.~~

Plan 2:
I implemented the skin selector thing and also added Muf as a second friend to CatDog. I think this is a better compromise than renaming or scraping the idea.

_(also linter will surely trigger now)_

### CODE REVIEW

**Comment by awesomekling** [comment]
No disrespect to IRL Muf, who I'm sure is a lovely cat, but I think we've all grown attached to the goofy `CatDog` name at this point and I think we should keep it. :)

**Comment by timmot** [comment]
Maybe some of CatDog's messages could be Muf inspired?

"Hello friend! Did you know my name is actually Muf?"

**Comment by IdanHo** [comment]
Looks like youre missing a couple of the textures for the other cat?

**Comment by TheMorc** [comment]
> Looks like youre missing a couple of the textures for the other cat?

Noticed it right now, it should be OK now

File: Userland/Demos/CatDog/CatDog.h [hunk 1]
```diff
@@ -23,19 +23,25 @@ class CatDog final : public GUI::Widget {
...
+    {
+        m_skin_name = skin_name;
+        m_skin = *Gfx::Bitmap::load_from_file(skin_path);
+    };
     void start_the_timer() { m_timer.start(); }
 
     Function<void()> on_click;
     Function<void(GUI::ContextMenuEvent&)> on_context_menu_request;
 
+    String get_skin_name() { return m_skin_name; }
```
Comment (linusg) [inline Userland/Demos/CatDog/CatDog.h:36]: Please don't prefix getters with `get_` as per the [code style guidelines](https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md):
 
>  Precede setters with the word "set". Use bare words for getters. Setter and getter names should match the names of the variables being set/gotten.

Also, no need for a string copy.

```suggestion
    String const& skin_name() { return m_skin_name; }
```

File: Userland/Demos/CatDog/CatDog.h [hunk 2]
```diff
@@ -44,33 +50,14 @@ class CatDog final : public GUI::Widget {
     Gfx::IntPoint m_temp_pos;
     Core::ElapsedTimer m_timer;
 
-    int m_curr_frame = 1;
+    int m_curr_frame, m_curr_second = 0;
```
Comment (linusg) [inline Userland/Demos/CatDog/CatDog.h:53]: Please don't declare multiple variables on the same line.

```suggestion
    int m_curr_frame = 0;
    int m_curr_second = 0;
```

Also, generally abbreviations should be avoided, unless they're more clear/common than the word itself (which isn't the case for "current").

File: Userland/Demos/CatDog/main.cpp [hunk 3]
```diff
@@ -6,15 +6,29 @@
...
 
+struct SkinMetadata {
+    String name;
+    String path;
+};
+
+Vector<SkinMetadata> g_skins;
+RefPtr<GUI::Menu> g_skins_menu;
+GUI::ActionGroup g_skins_group;
+
```
Comment (linusg) [inline Userland/Demos/CatDog/main.cpp:31]: Why are these global?

File: Userland/Demos/CatDog/main.cpp [hunk 4]
```diff
@@ -54,6 +68,39 @@ int main(int argc, char** argv)
...
+    auto current_skin_name = catdog_widget.get_skin_name();
+
+    {
+        int skin_identifier = 0;
+        for (auto& skin : g_skins) {
+            auto action = GUI::Action::create_checkable(skin.name, [&, skin_identifier](auto&) {
+                auto& skin = g_skins[skin_identifier];
+                dbgln("Skin switched to {} at path {}", skin.name, skin.path);
+                catdog_widget.set_skin(skin.name, skin.path);
```
Comment (linusg) [inline Userland/Demos/CatDog/main.cpp:95]: Perhaps this should take a `SkinMetadata` instead? Seems odd to group the values in a struct here but not elsewhere.

**Review by linusg — CHANGES_REQUESTED** [review]

File: Userland/Demos/CatDog/main.cpp (hunk 4 above)
Comment (TheMorc) [reply Userland/Demos/CatDog/main.cpp:95]: I based this part of the code from the startmenu's themes picker, so it is more less the same as that one.

**Comment by tomuta** [comment]
I know it's not necessarily part of the scope of this, but can't we change it to use one image with sprites? Seems quite inefficient and a lot of files just for one demo app ;)

**Pull request closed without merging**